### ISSUE TITLE
Dynamic query definition from the database mapping types

### ISSUE THREAD

**Issue by Smert** [juanluispaz/ts-sql-query#100]
Hello! How to get `TABLE_OR_VIEW` of select query? Now I can get only `[requiredTableOrView]` from `SelectExpressionBase`. I make a specific builder of CRUD interface around ts-sql-query.

My very simplified example of usage:
```typescript
builder() // My CRUD builder
  .withSelect((connection) => connection.selectFrom(tPost))
  .withField('name', () => tPost.name) // Types: OK
  .withField('email', () => tUser.name) // Types: ERROR, because no tUser in `withSelect`
  .withFilter('in', (value) => tPost.id.in(value)) // Types: OK 
  .withFilter('eq', (value) => tUser.id.eq(value)) // Types: ERROR, because no tUser in `withSelect`
```

I can check types of ValueSources in withField/withFilter using `TABLE_OR_VIEW`.

TypeScript `infer` gives me `Type instantiation is excessively deep and possibly infinite.`.

Maybe there can be [tableOrView] symbol?

**Comment by juanluispaz** [comment]
Hi, 

Can you explain more about what you are trying to do with this builder?

ts-sql-query already offers a lot of possibilities to deal with [dynamic queries](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/), some of them allow you to deal with extremely dynamic queries:

1.- You can build your query's where [condition dynamically](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#select-using-a-dynamic-filter) from an object representation "à la graphql"; that will give you a way to build it dynamically, having access to operations like `equals`or `in`

2.- You can use [`dynamicPick`](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#select-dynamically-picking-columns) to select make the query only returns the desired columns

3.- You can use `orderByFromString` or `orderByFromStringIfValue` to set the order by when it is provided outside the query (like in your Rest endpoint) in a safe way

4.- You can use `limit`, `limitIfValue`, `offset` and `offsetIfValue` if the page limit and offset are coming from outside the query.

You can see a complete example [here](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#select-using-a-dynamic-filter).

The main idea with ts-sql-query is to offer all you need to deal with complex dynamic queries. Be aware in the [advanced usage](https://ts-sql-query.readthedocs.io/en/stable/advanced-usage/) page; there is more information to [pass tables or views as parameters](https://ts-sql-query.readthedocs.io/en/stable/advanced-usage/#passing-tables-and-views-as-parameter) in function, or [extract all columns](https://ts-sql-query.readthedocs.io/en/stable/advanced-usage/#extract-columns) from a table or view, allowing you to do things like emulate a select *; as well there are several [utility types](https://ts-sql-query.readthedocs.io/en/stable/advanced-usage/#utility-types) to go even further.

I don't think you need a builder on top of ts-sql-query; maybe you are looking for something different I'm not getting right now; If you want, explain what you are trying to archive, and I can add any missing functionality to reach there.

**Note**: I already have an undocumented experimental variation of `dynamicPick` called `dynamicPickPaths` that receives an array with the columns names (or path) instead of an object with true value when the column is desired.

**Important**: ts-sql-query objects are complex due to the way SQL works; if you do `tPost.name`, the result is not a string type, nor a column of string; it is a column of type string in the table tPost plus other information like it is writable, it is coming from a left join, etc. In SQL every element has a context that you must keep in mind.

**Comment by juanluispaz** [comment]
Oh, 

I forgot to mention the [optional joins](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#optional-joins) and the [restrict access to values](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#restrict-access-to-values) that can be useful in extremely dynamic queries.

**Comment by Smert** [comment]
My builder creates typical objects for my UI like this:

```typescript
interface Entity<KEY, RECORD, EDITABLE_FIELD extends keyof RECORD & string> {
  [keySymbol]: KEY;
  [recordSymbol]: RECORD;
  [editableFieldSymbol]: EDITABLE_FIELD;
  
  getRecords<
    FIELD extends keyof RECORD & string
  >(ids: KEY[], fields: FIELD[]): Promise<
    Map<KEY, Pick<RECORD, FIELD>>
  >;
  updateRecords(changes: Map<KEY, Partial<Pick<RECORD, EDITABLE_FIELD>>>): Promise<void>;
}
```

I want to create a builder:
```typescript
const postEntity = builder()
  .withSelectQuery((connection: DBConnection) => {
    return connection.selectFrom(tPost)
      .optionalJoin(tUser).on(tUser.id.equals(tPost.userId));
  })
  .withUpdateQuery((connection: DBConnection) => {
    return connection.update(tPost)
      .optionalJoin(tUser).on(tUser.id.equals(tPost.userId));
  })
  .withField('id', {
    toSql() => tPost.id // Need to check this field in select query
  })
  .withField('name', {
    toSql() => tPost.name, // Need to check this field in select query
    editable: true // Need to check this field in update query
  })
  .withField('userName', {
    toSql() => tUser.name, // Need to check this field in select query
    editable: true // Need to check this field in update query
  })
  .withField('testSubQuery', (connection: DBConnection) => {
    return connection.selectFrom(...).selectOneColumn(connection.countAll());
  })
  .withPrimaryKey(['id'] as const)
  .build();
```

This builder generates entity and full type of Record like:
```typescript
  type PostEntity = typeof postEntity;
  type PostId = PostEntity[typeof keySymbol]; // number because tPost.id is number ValueSource
  type Post = PostEntity[typeof recordSymbol]; // {id: number, name: string, userName: string, testSubQuery: number}
  type PostEditableField = PostEntity[typeof editableFieldSymbol]; // 'name' | 'userName'
```

---

This builder has a lot of abilities and it allows me to describe the Entity in a very simple way and generate types.

**Comment by juanluispaz** [comment]
Let me see if I understand correctly; you want to have this object (I'm omitting generic types for now):

```ts
interface Entity {
    getRecords(ids: PK_TYPE[], fields: AVALIABLE_FIELD_TO_QUERY[]): Map< PK_TYPE, OBJECT_WITH_REQUESTED_FIELDS>
    updateRecords(changes: Map< PK_TYPE, OBJECT_WITH_SOME_FIELDS_TO_UPDATE): Promise<void>;
}
```

Is that what you are looking for? Can I offer you a different builder syntax if that is the case?

**Comment by Smert** [comment]
Exactly

**Comment by Smert** [comment]
Method `withSelectQuery` gets `DBConnection` type from generic:

```typescript
withSelectQuery<
  TConnection extends AbstractConnection<AnyDB>,
  TSelectQuery extends ... // All select query types
>(createSelectQuery: (connection: TConnection) => TSelectQuery);
```

`withSelectQuery` defines Connection type for another methods of the builder. This method should restrict TABLE_OR_VIEW for `withField`.

**Comment by juanluispaz** [comment]
I'm going to be adding several comments regarding this in several days. I will be posting different possibilities, getting closer to what I think you are looking for each time.

I'm making some changes (not published yet) to ensure are utility types are there and the documentation reflects the information as well; in the past, some additional function was added as experimental (then not included in the documentation).

The idea behind ts-sql-query is to be a query builder, including the most extreme dynamic query you can imagine, allowing you to keep control over the query in a type-safe way. 

I understand the idea of your builder to deal with the dynamic field selection in your query, but for that purpose, ts-sql-query already offers [Select dynamically picking columns](https://ts-sql-query.readthedocs.io/en/latest/queries/dynamic-queries/#select-dynamically-picking-columns), you don't need to assemble your query select fields, ts-sql-query does it for you.

Using [Select dynamically picking columns](https://ts-sql-query.readthedocs.io/en/latest/queries/dynamic-queries/#select-dynamically-picking-columns) it solves as well how to provide additional fields to the list of available fields (like `testSubQuery` that you put in your example); keep in mind if that field is not picked, the corresponding field will not be included in the sql query (allowing to have the subquery only when is required in case of your example. This approach has another additional value; now, you can get all the information in a single query instead of doing multiple calls to the backend, as I infer it will be happening in the builder you provided.

Be aware there are several additional features that you can take advantage of:
- [Select using a dynamic filter](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#select-using-a-dynamic-filter) that allows the caller to specify a condition that you can use in the where dynamically.
- [Optional joins](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#optional-joins) that will allow you to provide fiends in your entity that are coming from other tables; then in your query, you can implement a join (as optional join), that only will be generated in your query when is required
- [Restrict access to values](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#restrict-access-to-values) that allows you to restrict the access in some columns

Let's put an example (I will provide more in the future):

```ts
import { DynamicCondition, DynamicPickPaths, dynamicPickPaths, expandTypeFromDynamicPickPaths, PickValuesPath } from "ts-sql-query/dynamicCondition" 

class CustomerService {
    private connection
    private companyFields
    private companyWritableFields

    constructor(connection: DBConnection) {
        this.connection = connection
        this.companyFields = {
            ...extractColumnsFrom(tCompany),
            customerCount: connection
                .subSelectUsing(tCompany)
                .from(tCustomer)
                .where(tCustomer.companyId.equals(tCompany.id))
                .selectCountAll()
                .forUseAsInlineQueryValue()
        }

        this.companyWritableFields = {
            ...extractWritableColumnsFrom(tCompany, ['id']),
        }
    }


    async getCompaniesById<FIELDS extends DynamicPickPaths<typeof this.companyFields>>(ids: number[], fields: FIELDS[]): Promise<Map<number, PickValuesPath<typeof this.companyFields, FIELDS | 'id'>>> {
        const companies = await this.connection
            .selectFrom(tCompany)
            .where(tCompany.id.in(ids))
            .select(dynamicPickPaths(this.companyFields, fields, ['id']))
            .executeSelectMany()
        
        const result = new Map<number, PickValuesPath<typeof this.companyFields, FIELDS | 'id'>>()
        for (const company of companies) {
            result.set(company.id, expandTypeFromDynamicPickPaths(this.companyFields, fields, company, ['id']))
        }

        return result
    }

    async getCompanies<FIELDS extends DynamicPickPaths<typeof this.companyFields>>(fields: FIELDS[], condition: DynamicCondition<typeof this.companyFields>): Promise<PickValuesPath<typeof this.companyFields, FIELDS | 'id'>[]> {
        const companies = await this.connection
            .selectFrom(tCompany)
            .where(this.connection.dynamicConditionFor(this.companyFields).withValues(condition))
            .select(dynamicPickPaths(this.companyFields, fields, ['id']))
            .executeSelectMany()

        return expandTypeFromDynamicPickPaths(this.companyFields, fields, companies, ['id'])
    }

    async updateCompaniesById(changes: Map<number, UpdatableValues<typeof this.companyWritableFields>>): Promise<void> {
        for (const [id, change] of changes) {
            await this.connection
                .update(tCompany)
                .set(change)
                .where(tCompany.id.equals(id))
                .executeUpdate()
        }
    }
}
```

I will post another implementation with a focus on improving the type's  readability

**Comment by juanluispaz** [comment]
In this version, I improve the following:

- Reusable types that increase the readability of the code and allow one to take advantage of them in different places. Pay attention, when you query a company, the returning type only includes the requested properties plus the id.  
- Implemented as functional
- Update only allows the fields we are limiting to (in the previous version, I didn't include that rule)

```ts
export type CompanyFields = DynamicPickPaths<ReturnType<typeof getCompanyFields>>
export type CompanyInformation<FIELDS extends CompanyFields> = PickValuesPath<ReturnType<typeof getCompanyFields>, FIELDS | 'id'>
export type CompanyDynamicCondition = DynamicCondition<ReturnType<typeof getCompanyFields>>
export type UpdatableCompany = UpdatableValues<ReturnType<typeof getCompanyWritableFields>>

export async function getCompaniesById<FIELDS extends CompanyFields>(connection: DBConnection, ids: number[], fields: FIELDS[]): Promise<Map<number, CompanyInformation<FIELDS>>> {
    const companyFields = getCompanyFields(connection)

    const companies = await connection
        .selectFrom(tCompany)
        .where(tCompany.id.in(ids))
        .select(dynamicPickPaths(companyFields, fields, ['id']))
        .executeSelectMany()
    
    const result = new Map<number, CompanyInformation<FIELDS>>()
    for (const company of companies) {
        result.set(company.id, expandTypeFromDynamicPickPaths(companyFields, fields, company, ['id']))
    }

    return result
}

export async function getCompanies<FIELDS extends CompanyFields>(connection: DBConnection, fields: FIELDS[], condition: CompanyDynamicCondition): Promise<CompanyInformation<FIELDS>[]> {
    const companyFields = getCompanyFields(connection)

    const companies = await connection
        .selectFrom(tCompany)
        .where(connection.dynamicConditionFor(companyFields).withValues(condition))
        .select(dynamicPickPaths(companyFields, fields, ['id']))
        .executeSelectMany()

    return expandTypeFromDynamicPickPaths(companyFields, fields, companies, ['id'])
}

export async function updateCompaniesById(connection: DBConnection, changes: Map<number, UpdatableCompany>): Promise<void> {
    const allowedColumns = getCompanyWritableFields(connection)

    for (const [id, change] of changes) {
        const valuesToUpdate = getOnlyAllowedValues(allowedColumns, change)

        await connection
            .update(tCompany)
            .set(valuesToUpdate)
            .where(tCompany.id.equals(id))
            .executeUpdate()
    }
}

export async function updateCompanyById(connection: DBConnection, id: number, changes: UpdatableCompany): Promise<void> {
    const allowedColumns = getCompanyWritableFields(connection)
    const valuesToUpdate = getOnlyAllowedValues(allowedColumns, changes)

    await connection
        .update(tCompany)
        .set(valuesToUpdate)
        .where(tCompany.id.equals(id))
        .executeUpdate()
}

function getCompanyFields(connection: DBConnection) {
    return {
        ...extractColumnsFrom(tCompany),
        customerCount: connection
            .subSelectUsing(tCompany)
            .from(tCustomer)
            .where(tCustomer.companyId.equals(tCompany.id))
            .selectCountAll()
            .forUseAsInlineQueryValue()
    }
}

function getCompanyWritableFields(_connection: DBConnection) {
    return extractWritableColumnsFrom(tCompany, ['id'])
}

/*
 * This function ensure only the values we are allowing to update are included in the final update
 */
function getOnlyAllowedValues<TYPE, TARGET extends {[P in keyof TYPE]?: any}>(type: TYPE, target: TARGET): {[P in keyof TYPE]: TARGET[P]} {
    const result : {[P in keyof TYPE]: TARGET[P]} = {} as any
    for (const prop in type) {
        result[prop] = target[prop]
    }
    return result
}
```

**Comment by juanluispaz** [comment]
I add a new change; using update shape. The shape functionality was an experimental feature that I'm publishing for general usage in the next release:

```ts
export type CompanyFields = DynamicPickPaths<ReturnType<typeof getCompanyFields>>
export type CompanyInformation<FIELDS extends CompanyFields> = PickValuesPath<ReturnType<typeof getCompanyFields>, FIELDS | 'id'>
export type CompanyDynamicCondition = DynamicCondition<ReturnType<typeof getCompanyFields>>
const updateCompanyShape = extractWritableShapeFrom(tCompany, ['id'])
export type UpdatableCompany = UpdatableValuesShapedAs<typeof tCompany, typeof updateCompanyShape>

export async function getCompaniesById<FIELDS extends CompanyFields>(connection: DBConnection, ids: number[], fields: FIELDS[]): Promise<Map<number, CompanyInformation<FIELDS>>> {
    const companyFields = getCompanyFields(connection)

    const companies = await connection
        .selectFrom(tCompany)
        .where(tCompany.id.in(ids))
        .select(dynamicPickPaths(companyFields, fields, ['id']))
        .executeSelectMany()
    
    const result = new Map<number, CompanyInformation<FIELDS>>()
    for (const company of companies) {
        result.set(company.id, expandTypeFromDynamicPickPaths(companyFields, fields, company, ['id']))
    }

    return result
}

export async function getCompanies<FIELDS extends CompanyFields>(connection: DBConnection, fields: FIELDS[], condition: CompanyDynamicCondition): Promise<CompanyInformation<FIELDS>[]> {
    const companyFields = getCompanyFields(connection)

    const companies = await connection
        .selectFrom(tCompany)
        .where(connection.dynamicConditionFor(companyFields).withValues(condition))
        .select(dynamicPickPaths(companyFields, fields, ['id']))
        .executeSelectMany()

    return expandTypeFromDynamicPickPaths(companyFields, fields, companies, ['id'])
}

export async function updateCompaniesById(connection: DBConnection, changes: Map<number, UpdatableCompany>): Promise<void> {
    for (const [id, change] of changes) {
        await connection
            .update(tCompany)
            .shapedAs(updateCompanyShape)
            .set(change)
            .where(tCompany.id.equals(id))
            .executeUpdate(1, 1) // Validate if the return of the query indicates min 1 record and max 1 record is updated (no reverted in the DB, use transactions for this)
    }
}

export async function updateCompanyById(connection: DBConnection, id: number, changes: UpdatableCompany): Promise<void> {
    await connection
        .update(tCompany)
        .shapedAs(updateCompanyShape)
        .set(changes)
        .where(tCompany.id.equals(id))
        .executeUpdate(1, 1) // Validate if the return of the query indicates min 1 record and max 1 record is updated (no reverted in the DB, use transactions for this)
}

function getCompanyFields(connection: DBConnection) {
    return {
        ...extractColumnsFrom(tCompany),
        customerCount: connection
            .subSelectUsing(tCompany)
            .from(tCustomer)
            .where(tCustomer.companyId.equals(tCompany.id))
            .selectCountAll()
            .forUseAsInlineQueryValue()
    }
}
```

**Comment by juanluispaz** [comment]
Released ts-sql-query with the changes mentioned here. In this version, I include several utility types to allow you to build your own types using the database definition and document more the functionalities required here; additionally, the experimental features have been published and documented.

Let me know if this approach works for you or if I am missing something.

**Comment by Smert** [comment]
Thank you for examples! It looks interesting.

I want to describe filter types in my builder. `dynamicConditionFor` doesn't allow to define `WHERE column IN (SELECT ...)` or `WHERE EXISTS (SELECT ...)`.

```typescript
Builder.create()
  .withSelect((connection: DBConnection) => connection.selectFrom(tPost))
  .withFilter('category', (value: number) => {
    const select = connection.selectFrom(tPostCategory) // many to many table postId-categoryId
      .where(tPostCategory.categoryId.equals(value))
      .selectOneColumn(tPostCategory.postId);
    return tPost.id.in(select); // TS: tPost should be in withSelect query
  });
  
// Simplified builder
class Builder<
  TTableOrViewRef extends ITableOrViewRef<AnyDB>,
  TFilter
> {
  withSelectQuery<
    TConnection extends AbstractConnection<AnyDB>,
    TSelectQuery extends ... // It's hard to describe SelectQuery<TConnection[typeof database]> but I can put there all allowed select types
  >(
    createSelectQuery: (connection: TConnection) => TSelectQuery
  ): Builder<
    ExtractTableOrViewRef<TSelectQuery>, // I don't have variants to get TableOrViewRef from the query 😢
    TFilter
  >;

  withFilter<TFilterName extends string, TFilterType>(
    filterName: TFilterName,
    (value: TFilterType) => BooleanValueSource<TTableOrViewRef, OptionalType>
  ): Builder<
    TTableOrViewRef,
    TFilter | [TFilterName, TFilterType]
  >;

   // ...
}
```

I think that `dynamicConditionFor` can be useful for simple variants. With complex filters I should have my own filter to `ValueSource` transformers

**Comment by juanluispaz** [comment]
mmmm, interesting

In the fields to use in the condition you can use any field or expression available in your query that represents a value:

```ts
const selectFields = {
        id: tCustomer.id,
        firstName: tCustomer.firstName,
        lastName: tCustomer.lastName,
        birthday: tCustomer.birthday,
        companyName: tCompany.name, // This from other table, can be any expression
    }
    
    const fieldsForcondition = {
        ...selectFields,
        companyFirstLetter: tCompany.name.substrToEnd(0) // An expression
    }

    type FilterType = DynamicCondition<typeof fieldsForcondition>
    
    const filter: FilterType = {
        or: [
            { firstName: { startsWithInsensitive: 'John' } },
            { lastName: { startsWithInsensitiveIfValue: 'Smi', endsWith: 'th' } }
        ],
        companyName: {equals: 'ACME'},
        companyFirstLetter: {equals: 'A'}
    }
    
    const dynamicWhere = connection.dynamicConditionFor(fieldsForcondition).withValues(filter)
    
    const customersWithDynamicCondition = await connection.selectFrom(tCustomer)
        .innerJoin(tCompany).on(tCustomer.companyId.equals(tCompany.id))
        .where(dynamicWhere)
        .select(selectFields)
        .orderBy('firstName', 'insensitive')
        .orderBy('lastName', 'asc insensitive')
        .executeSelectMany()
```

But, there is no way to express a subquery expression; but I can add it; then you will have the possibility to include a subquery in `fieldsForcondition`.

What do you think?

**Comment by juanluispaz** [comment]
The type for the condition in this subqueries will be:

```ts
export interface SubqueryFilter<TYPE> extends Filter {
    inIfValue?: TYPE[] | null | undefined
    in?: TYPE[]
    notInIfValue?: TYPE[] | null | undefined
    notIn?: TYPE[]
    containsIfValue?: string | null | undefined
    contains?: TYPE
    notContainsIfValue?: TYPE | null | undefined
    notContains?: TYPE
    exists?: boolean
    notExists?: boolean
}
```

**Comment by juanluispaz** [comment]
I will add as well the possibility to extend the rules of a field or add a field with your own rules; this will allow you to add any required condition, no matter whether they are supported or not by default in ts-sql-query. This will allow you to add any complex rules you want.

**Comment by juanluispaz** [comment]
Released ts-sql-query 1.53.0 adding the possibility to you define your own rules in a dynamic condition. See a full example at [Full dynamic select](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#full-dynamic-select)

Let me know your feedback.
